pkg/auth/remoteauthenticator/authenticator.go

```diff
@@ -0,0 +1,169 @@
+	payload, err := json.Marshal(&AuthenticationRequest{Username: username, Password: password})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed marshaling request body: %w", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, ra.serviceURL, bytes.NewReader(payload))
+
+	if err != nil {
+		return nil, fmt.Errorf("failed creating request to remote authenticator: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	log = log.WithField("url", req.URL.String())
+
+	client := ra.client()
+
+	log.Debug("starting http request to remote authenticator")
+
+	resp, err := client.Do(req)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed sending request to remote authenticator: %w", err)
+	}
+
+	log = log.WithField("status_code", resp.StatusCode)
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("bad status code %d: %w", resp.StatusCode, auth.ErrUnexpectedStatusCode)
+	}
+
+	log.Debug("got response from remote authenticator")
+
+	defer resp.Body.Close()
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed reading response body: %w", err)
+	}
+	return body, nil
+}
+
+func (ra *RemoteAuthenticator) AuthenticateUser(ctx context.Context, username, password string) (string, error) {
+	log := ra.Logger.WithContext(ctx).WithField("input_username", username)
+
+	data, err := ra.doRequest(ctx, username, password, log)
+
+	if err != nil {
+		return "", fmt.Errorf("doing http request %s: %w", username, err)
+	}
+
+	var res AuthenticationResponse
+
+	if err := json.Unmarshal(data, &res); err != nil {
+		return "", fmt.Errorf("unmarshaling authenticator response %s: %w", username, err)
+	}
+
+	dbUsername := username
+
+	// if the external authentication service provided an external user identifier, use it as the username
+
+	if res.ExternalUserIdentifier != nil && *res.ExternalUserIdentifier != "" {
+		log = ra.Logger.WithField("external_user_identifier", *res.ExternalUserIdentifier)
+		dbUsername = *res.ExternalUserIdentifier
+	}
+
+	user, err := ra.AuthService.GetUser(ctx, dbUsername)
+
+	if err == nil {
+		log.WithField("user", fmt.Sprintf("%+v", user)).Debug("Got existing user")
+		return user.Username, nil
+	}
+	if !errors.Is(err, auth.ErrNotFound) {
+		return "", fmt.Errorf("get user %s: %w", dbUsername, err)
+	}
+
+	log.Info("could not get user; creating them")
+
+	newUser := &model.User{
+		CreatedAt:    time.Now().UTC(),
+		Username:     dbUsername,
+		FriendlyName: &username,
+		Source:       RemoteAuthSource,
+	}
+
+	_, err = ra.AuthService.CreateUser(ctx, newUser)
+	if err != nil {
+		return "", fmt.Errorf("create backing user for remote auth user %s: %w", newUser.Username, err)
+	}
+	_, err = ra.AuthService.CreateCredentials(ctx, newUser.Username)
+	if err != nil {
+		return "", fmt.Errorf("create credentials for remote auth user %s: %w", newUser.Username, err)
+	}
```

Comment:
why do we need to create credentials for that user?

Reply:
Actually it's a good question, right now it's backward compatible [with the LDAP implementation](https://github.com/treeverse/lakeFS/blob/master/pkg/auth/authenticator.go#L237). 

I couldn't find a reason for it, testing without this function `CreateCredentials` seems to work. 
Do you happen to know why it's here @arielshaqed  @nopcoder ?

Reply:
removed!